contrib/pam_ssh_agent_auth/pam-ssh-agent-auth.c

```diff
@@ -0,0 +1,195 @@
+#include "digest.h"
+#include "log.h"
+#include "misc.h"
+#include "sshbuf.h"
+#include "sshkey.h"
+#include "servconf.h"
+
+#define CHALLENGE_PREFIX	"pam-ssh-agent-auth-challenge"
+#define CHALLENGE_NONCE_LEN	32
+#define UNUSED(expr) do { (void)(expr); } while (0)
+
+char		*authorized_keys_file = "/etc/security/authorized_keys";
+
+void
+auth_debug_add(const char *fmt,...)
+{
+}
+void
+auth_log_authopts(const char *loc, const struct sshauthopt *opts, int do_remote)
+{
+}
+
+/* obtain a list of keys from the agent */
+static int
+pam_get_agent_identities(int *agent_fdp,
+    struct ssh_identitylist **idlistp)
+{
+	int r, agent_fd;
+	struct ssh_identitylist *idlist;
+
+	if ((r = ssh_get_authentication_socket(&agent_fd)) != 0) {
+		if (r != SSH_ERR_AGENT_NOT_PRESENT)
+			debug_fr(r, "ssh_get_authentication_socket");
+		return r;
+	}
+	if ((r = ssh_fetch_identitylist(agent_fd, &idlist)) != 0) {
+		debug_fr(r, "ssh_fetch_identitylist");
+		close(agent_fd);
+		return r;
+	}
+	/* success */
+	*agent_fdp = agent_fd;
+	*idlistp = idlist;
+	debug_f("agent returned %zu keys", idlist->nkeys);
+	return 0;
+}
+
+PAM_EXTERN int
+pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv)
+{
+	FILE *f = NULL;
+	struct sshbuf *sigbuf = NULL;
+	u_char *nonce;
+	struct sshauthopt *authoptsp = NULL;
+	struct ssh_identitylist *idlist = NULL;
+	int ret = PAM_AUTH_ERR, agent_fd = -1;
+	struct passwd *pw = getpwuid(0);
+	size_t i;
+
+	for(; argc > 0; ++argv, argc--) {
+		if(strncasecmp(*argv, "file=", strlen("file=")) == 0 ) {
+			authorized_keys_file = (char *) *argv + strlen("file=");
+		}
+	}
+
+	if (pam_get_agent_identities(&agent_fd, &idlist) != 0) {
+		pam_syslog(pamh, LOG_CRIT, "pam_get_agent_identities() failed.");
+		goto exit;
+	}
+
+	if ((f = auth_openkeyfile(authorized_keys_file, pw, 1)) == NULL) {
+		pam_syslog(pamh, LOG_CRIT, "authorized_keys open failed.");
+		goto exit;
+	}
+
+	for (i = 0; i < idlist->nkeys; i++) {
+		/* Check if key in authorized_keys */
+		if (!auth_check_authkeys_file(pw, f, authorized_keys_file,
+		    idlist->keys[i], NULL, NULL, &authoptsp))
+			continue;
+
+		/* Generate random challenge */
+		if ((sigbuf = sshbuf_new()) == NULL ||
+		    sshbuf_put_cstring(sigbuf, CHALLENGE_PREFIX) != 0 ||
+		    sshbuf_reserve(sigbuf, CHALLENGE_NONCE_LEN, &nonce) != 0)
+			goto exit;
+
+		arc4random_buf(nonce, CHALLENGE_NONCE_LEN);
+
+		/* Sign challenge via ssh-agent */
+		u_char *sig = NULL;
+		size_t	slen = 0;
+		if (ssh_agent_sign(agent_fd, idlist->keys[i], &sig, &slen,
+		    sshbuf_ptr(sigbuf), sshbuf_len(sigbuf), NULL, 0) != 0)
+			goto exit;
+
+		/* Verify signature */
+		if (sshkey_verify(idlist->keys[i], sig, slen,
+		    sshbuf_ptr(sigbuf), sshbuf_len(sigbuf),
+		    NULL, 0, NULL) == 0) {
+			pam_syslog(pamh, LOG_INFO, "Found matching %s key: %s",
+			    sshkey_type(idlist->keys[i]),
+			    sshkey_fingerprint(idlist->keys[i], SSH_DIGEST_SHA256,
+			    SSH_FP_DEFAULT));
+
+			ret = PAM_SUCCESS;
```

Comment:
strictly, there's a couple more checks that should be done here wrt FIDO keys and touch/PIN required flags. These will require you to keep the sig_details from the verify call and check them similarly to https://github.com/openssh/openssh-portable/blob/V_9_0_P1/auth2-pubkey.c#L232-L266

Reply:
Also as much of auth_authorise_keyopts() as you can, at the very least the valid-from/to stuff